cabal-install/Distribution/Client/CmdSdist.hs

```diff
@@ -0,0 +1,321 @@
+    , commandUsage = \pname ->
+        "Usage: " ++ pname ++ " new-sdist [FLAGS] [PACKAGES]\n"
+    , commandDescription  = Just $ \_ ->
+        "Generates tarballs of project packages suitable for upload to Hackage."
+    , commandNotes = Nothing
+    , commandDefaultFlags = defaultSdistFlags
+    , commandOptions = \showOrParseArgs ->
+        [ optionVerbosity
+            sdistVerbosity (\v flags -> flags { sdistVerbosity = v })
+        , optionDistPref
+            sdistDistDir (\dd flags -> flags { sdistDistDir = dd })
+            showOrParseArgs
+        , option [] ["project-file"]
+            "Set the name of the cabal.project file to search for in parent directories"
+            sdistProjectFile (\pf flags -> flags { sdistProjectFile = pf })
+            (reqArg "FILE" (succeedReadE Flag) flagToList)
+        , option ['l'] ["list-only"]
+            "Just list the sources, do not make a tarball"
+            sdistListSources (\v flags -> flags { sdistListSources = v })
+            trueArg
+        , option ['z'] ["null-sep"]
+            "Separate the source files with NUL bytes rather than newlines."
+            sdistNulSeparated (\v flags -> flags { sdistNulSeparated = v })
+            trueArg
+        , option [] ["archive-format"] 
+            "Choose what type of archive to create. No effect if given with '--list-only'"
+                sdistArchiveFormat (\v flags -> flags { sdistArchiveFormat = v })
+            (choiceOpt
+                [ (Flag TargzFormat, ([], ["targz"]),
+                        "Produce a '.tar.gz' format archive (default and required for uploading to hackage)")
+                , (Flag ZipFormat,   ([], ["zip"]),
+                        "Produce a '.zip' format archive")
+                ]
+            )
+        , option ['o'] ["output-dir"]
+            "Choose the output directory of this command. '-' sends all output to stdout"
+            sdistOutputPath (\o flags -> flags { sdistOutputPath = o })
+            (reqArg "PATH" (succeedReadE Flag) flagToList)
+        ]
+    }
+
+data SdistFlags = SdistFlags
+    { sdistVerbosity     :: Flag Verbosity
+    , sdistDistDir       :: Flag FilePath
+    , sdistProjectFile   :: Flag FilePath
+    , sdistListSources   :: Flag Bool
+    , sdistNulSeparated  :: Flag Bool
+    , sdistArchiveFormat :: Flag ArchiveFormat
+    , sdistOutputPath    :: Flag FilePath
+    }
+
+defaultSdistFlags :: SdistFlags
+defaultSdistFlags = SdistFlags
+    { sdistVerbosity     = toFlag normal
+    , sdistDistDir       = mempty
+    , sdistProjectFile   = mempty
+    , sdistListSources   = toFlag False
+    , sdistNulSeparated  = toFlag False
+    , sdistArchiveFormat = toFlag TargzFormat
+    , sdistOutputPath    = mempty
+    }
+
+--
+
+sdistAction :: SdistFlags -> [String] -> GlobalFlags -> IO ()
+sdistAction SdistFlags{..} targetStrings globalFlags = do
+    let verbosity = fromFlagOrDefault normal sdistVerbosity
```

Comment:
If you use Emacs, you can make formatting prettier here (and elsewhere) with `M-x align-regexp =`. If you don't, don't bother.